src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb

```diff
@@ -81,6 +81,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         Private ReadOnly _syntaxTrees As ImmutableArray(Of SyntaxTree)
 
         ''' <summary>
+        ''' The syntax trees explicitly given to the compilation at creation, in set form.
+        ''' </summary>
+        Private ReadOnly _syntaxTreeSet As HashSet(Of SyntaxTree)
```

Comment:
Adding a new HashSet for this purpose feels like an overkill. A SyntaxTree knows what kind of tree it is, there is an extension method IsEmbeddedOrMyTemplateTree that can be used to check if it is one of the compiler added trees and this method is likely to be faster than a HashSet lookup.